packages/editor/src/lib/exports/StyleEmbedder.ts

```diff
@@ -0,0 +1,310 @@
+		return await this.fonts.createCss()
+	}
+
+	dispose() {
+		destroyDefaultStyleFrame()
+	}
+}
+
+interface ReadStyleOpts {
+	defaultStyles: ReadonlyStyles
+	parentStyles: ReadonlyStyles
+}
+
+function styleFromElement(element: Element, { defaultStyles, parentStyles }: ReadStyleOpts) {
+	// `computedStyleMap` produces a more accurate representation of the styles, but it's not
+	// supported in firefox at the time of writing. So we fall back to `getComputedStyle` if it's
+	// not available.
+	if (element.computedStyleMap) {
+		return styleFromComputedStyleMap(element.computedStyleMap(), { defaultStyles, parentStyles })
+	}
+	return styleFromComputedStyle(getComputedStyle(element), { defaultStyles, parentStyles })
+}
+
+function styleFromPseudoElement(element: Element, pseudo: string) {
+	// the equivalent of `computedStyleMap` for pseudo-elements isn't even fully specced out yet, so
+	// for those we have to use `getComputedStyle` in all browsers.
+	const style = getComputedStyle(element, pseudo)
+
+	const content = style.getPropertyValue('content')
+	if (content === '' || content === 'none') {
+		return undefined
+	}
+
+	return styleFromComputedStyle(style, { defaultStyles: NO_STYLES, parentStyles: NO_STYLES })
+}
+
+function styleFromComputedStyleMap(
+	style: StylePropertyMapReadOnly,
+	{ defaultStyles, parentStyles }: ReadStyleOpts
+) {
+	const styles: Record<string, string> = {}
+	for (const property of style.keys()) {
+		if (!shouldIncludeCssProperty(property)) continue
+
+		const value = style.get(property)!.toString()
+
+		if (defaultStyles[property] === value) continue
+		if (parentStyles[property] === value && isPropertyInherited(property)) continue
+
+		styles[property] = value
+	}
+
+	return styles
+}
+
+function styleFromComputedStyle(
+	style: CSSStyleDeclaration,
+	{ defaultStyles, parentStyles }: ReadStyleOpts
+) {
+	const styles: Record<string, string> = {}
+	for (const property of style) {
+		if (!shouldIncludeCssProperty(property)) continue
+
+		const value = style.getPropertyValue(property)
+
+		if (defaultStyles[property] === value) continue
+		if (parentStyles[property] === value && isPropertyInherited(property)) continue
+
+		styles[property] = value
+	}
+	return styles
+}
+
+function formatCss(style: ReadonlyStyles) {
+	let cssText = ''
+	for (const [property, value] of Object.entries(style)) {
+		cssText += `${property}: ${value};`
+	}
+	return cssText
+}
+
+// when we're figuring out the default values for a tag, we need read them from a separate document
+// so they're not affected by the current document's styles
+let defaultStyleFrame:
+	| { iframe: HTMLIFrameElement; foreignObject: SVGForeignObjectElement; document: Document }
+	| undefined
+const defaultStylesByTagName: Record<string, ReadonlyStyles> = {}
+function getDefaultStyleFrame() {
+	if (!defaultStyleFrame) {
+		const frame = document.createElement('iframe')
+		frame.inert = true
+		frame.style.display = 'none'
+		document.body.appendChild(frame)
```

Comment:
out of curiosity, is it strictly necessary to add to the document?
can something like `DocumentFragment` be used? https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment

Reply:
yes - `getComputedStyle`/`computedStyleMap` needs the element to be in the DOM to have the right values & layout applied

Reply:
with all you shadow dom work, it makes me think we should be putting this in a shadow node heh. 

it's fine though but part of me is worrying about pollution of the rest of the dom potentially (if something isn't buttoned up either via this PR or future tweaks).

(actually, side-note, i wonder if we should have a "sandbox" node to add temporary nodes like this. since tldraw is meant to be integrated into a larger app, it feels like we should maybe stay within a generally `inert` div vs. adding it to the top-level of the `<body>` element)